Add Radiant MLHub (REF) Cloud Cover Dataset

KennSmithDS opened the pull request:
This PR adds the [Cloud Cover Detection Challenge ](https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1) dataset which was generated as part of a [crowdsourcing competition](https://www.drivendata.org/competitions/83/cloud-cover/) as hosted on the Radiant MLHub.

Dataset features:
* Sentinel-2 L2A (RGB and NIR bands only) 
* 22,728 chips in the training dataset, and 10,980 chips in the test dataset
* Chips represent imagery between 2018 and 2020
* GeoTiff labels represent binary classification if a pixel is labelled as a cloud or not

To Dos:
* ~~Resolve linter errors~~
* Create new sample data with `data.py` script

calebrob6 commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,410 @@
+            AssertionError: if ``bands`` is not a tuple
+            ValueError: if an invalid band name is provided
+        """
+        assert isinstance(bands, List), "The list of bands must be a tuple"
```
mypy should handle type checking this. That said, does `bands` need to be a List, or can it be a more general `Sequence`?

KennSmithDS replied:
I have removed the assertion statement, and changed typing annotations for bands to `Sequence[str]`

KennSmithDS commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,410 @@
+        return target_tensor
+
+    @staticmethod
+    def _validate_stac_object(object_path: str) -> bool:
```
That was the intended purposes of the method. If the STAC object doesn't exist, would it be better to let Python native library's error handling raise these instead of the `RuntimeError` raised? In other words, do you believe this method should simply be removed?

KennSmithDS commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,410 @@
+
+            if self._validate_stac_object(label_asset_path):
+                source_item_hrefs = [
+                    os.path.join(self.root, link["href"].replace("../../", ""))
```
Looks like the cleanest way to do this is `os.path.abspath`. I will push that change in a commit.

KennSmithDS replied:
Update on this, I actually ran into issues using `os.path.abspath`, `os.path.realpath` and `Path().absolute()`. None of these build the appropriate file paths because the path provided in the `stac.json` file for a STAC Item is relative to the collection, e.g.:

> ../../ref_cloud_cover_detection_challenge_v1_test_source/ref_cloud_cover_detection_challenge_v1_test_source_aaaa/stac.json

I'm not sure what the best way to do this is without calling `.replace('../../', '')`

adamjstewart commented on tests/data/ref_cloud_cover_detection_challenge_v1/ref_cloud_cover_detection_challenge_v1_test_labels 2/collection.json:
```diff
@@ -0,0 +1,66 @@
+{
```
All of these `.DS_Store` files can be removed from the PR

adamjstewart commented on tests/datasets/test_cloud_cover.py:
```diff
@@ -0,0 +1,118 @@
+
+from torchgeo.datasets import CloudCoverDetection
+
+# TODO use with pytest.fixture params
+# from requests import request
+# from _pytest.fixtures import SubRequest
+# from torch.utils.data import ConcatDataset
```
Commented out code should be removed prior to merge

adamjstewart commented on tests/datasets/test_cloud_cover.py:
```diff
@@ -0,0 +1,118 @@
+
+        root = str(tmp_path)
+        split = "test"  # request.param
+        transforms = nn.Identity()  # type: ignore[no-untyped-call]
```
```suggestion
        transforms = nn.Identity()
```
Latest version of PyTorch adds type hints for this

adamjstewart commented on torchgeo/datasets/__init__.py:
```diff
@@ -115,6 +116,7 @@
     "ChesapeakeVA",
     "ChesapeakeWV",
     "ChesapeakeCVPR",
+    "CloudCoverDetection",
```
Can you also add a section to `docs/api/datasets.rst`? Should go under the non-geospatial datasets section.

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+    """Cloud Cover Detection Challenge dataset.
+
+    This training dataset was generated as part of a
+    [crowdsourcing competition](https://www.drivendata.org/competitions/83/cloud-cover/)
```
```suggestion
    `crowdsourcing competition <https://www.drivendata.org/competitions/83/cloud-cover/>`_
```
Docs use rST, not markdown

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+    This training dataset was generated as part of a
+    [crowdsourcing competition](https://www.drivendata.org/competitions/83/cloud-cover/)
+    on DrivenData.org, and later on was validated using a team of expert annotators. See
+    `this website<https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1>`
```
```suggestion
    `this website <https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1>`_
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+
+    Each chip has:
+
+    * 4 multi-spectral bands from Sentinel-2 L2A product. The four bands are
+    [B02, B03, B04, B08] (refer to Sentinel-2 documentation for more
+    information about the bands).
+    * Label raster for the corresponding source tile representing a binary classifcation
+    for if the pixel is is a cloud or not.
```
```suggestion
    * 4 multi-spectral bands from Sentinel-2 L2A product. The four bands are
      [B02, B03, B04, B08] (refer to Sentinel-2 documentation for more
      information about the bands).
    * Label raster for the corresponding source tile representing a binary classifcation
      for if the pixel is is a cloud or not.
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+       This dataset requires the following additional library to be installed:
+
+       * `radiant-mlhub <https://pypi.org/project/radiant-mlhub/>`_ to download the
+         imagery and labels from the Radiant Earth MLHub
```
You should add a line like:
```rst
    .. versionadded:: 0.3                                                                
```
to the end of this docstring. This documents which version of TorchGeo added this dataset.

Just a heads up, we're planning a 0.3.0 release in the next few days. If this is finished after 0.3 is released, just change this to 0.4.

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+    def __init__(
+        self,
+        root: str = "data",
+        chip_size: int = 512,
```
This parameter isn't used anywhere and can be removed

adamjstewart reviewed (changes requested):
Still needs a `data.py` script to generate the fake data. It also looks like some of the tests are failing. You'll need to rebase/merge on top of main to run some of the new required tests. Let me know if I can help with any of this.

KennSmithDS commented:
@adamjstewart @calebrob6 thank you for the feedback and recommendations. I've rebased/merged the latest release with my local branch, and fixed the failing tests, as well as implemented the suggestions above. The only thing I still have questions on is the `data.py` script.

Is this requirement for contributing documented anywhere in the repository? I've been looking at a few different examples to get a sense for what it does, e.g. `agb_live_woody_density`, `naip`, and `baywide` datasets. Each have their own flavor, but the common workflow I see is defining the profile information of a raster file to be written using rasterio. Questions this raises for me are:

- Is it required to use the long form WKT string for the `profile["crs"]` attribute or can you simply pass in the simpler versions, e.g. `CRS.from_epsg(4326)` or `CRS.from_string("EPSG:4326")`?
- Is it necessary/recommended to use a colormap for the geotiffs written to disk?
- Should there be a min/max value for the random values generated for the rasterio dataset writer class output? The above examples in the repo just create a random image np array of any dtype value

adamjstewart commented:
> Is this requirement for contributing documented anywhere in the repository?

The fact that it's needed is documented at https://torchgeo.readthedocs.io/en/latest/user/contributing.html#datasets. Some tips for writing the script can be found at https://github.com/microsoft/torchgeo/blob/main/tests/data/README.md

For all of your other questions, the answer is that you should try to match the real data from your dataset. 

* If your dataset contains raster images with geospatial metadata, you should create fake data using rasterio. If your dataset contains PNG or JPEG images with no geospatial metadata, you should create fake data using PIL. 
* For raster datasets, you should use the same CRS and dtype as the real files. 
* The random values can range from the min/max value of that dtype, which can be computed using `np.iinfo(dtype).min` and `np.iinfo(dtype).max` (or `np.ffinfo` for float dtypes). 
* If your dataset contains labels with a colormap, your fake data should to.

Hope that answers your questions!

KennSmithDS commented:
@adamjstewart I had missed the README file nestled under `tests/data/README.md` so thanks for pointing that out!

KennSmithDS commented:
@adamjstewart I'm assuming the answer to this question is yes, but to clarify, should the `data.py` script also generate the `collection.json` and `stac.json` files used by the method I added `[_load_collections()](https://github.com/KennSmithDS/torchgeo/blob/7a3d74fb0614da0973dbc2306513c3ce5312128f/torchgeo/datasets/cloud_cover.py#L250)`

adamjstewart commented:
Yep! At some point I'm planning to convert these data.py files to pytest fixtures to generate all of our data at test time instead of storing it in the repo. So if the data.py generates those files too that would help a lot.

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+from numpy.typing import NDArray
+from torch import Tensor
+
+from .geo import VisionDataset
+from .utils import check_integrity, download_radiant_mlhub_dataset, extract_archive
+
+
+# TODO: read geospatial information from stac.json files
+class CloudCoverDetection(VisionDataset):
```
```suggestion
from .geo import NonGeoDataset
from .utils import check_integrity, download_radiant_mlhub_dataset, extract_archive


# TODO: read geospatial information from stac.json files
class CloudCoverDetection(NonGeoDataset):
```
We renamed the base class.

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,390 @@
+import numpy as np
+import rasterio
+import torch
+from numpy.typing import NDArray
```
This module isn't available in older versions of numpy, which is why the tests are failing. If you instead remove this import and use `"np.typing.NDArray[...]"` (note the quotes) then the import is only evaluated when you run `mypy`, which we only run with the latest version of numpy.

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,389 @@
+        return target_tensor
+
+    @staticmethod
+    def _read_json_data(object_path: str) -> Any:
```
Can you add a docstring to this?

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,389 @@
+            bands: user-provided tuple of bands to load
+
+        Raises:
+            AssertionError: if ``bands`` is not a tuple
```
I don't see an assert statement in this, you can remove this line

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,389 @@
+        """
+        indexed_chips = []
+        label_collection = [
+            c for c in self.collection_names[self.split] if c.__contains__("label")
```
```suggestion
            c for c in self.collection_names[self.split] if "label" in c
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,389 @@
+        )
+        item_meta["target"] = [label_asset_path]
+
+        source_item_hrefs = [
+            os.path.join(self.root, link["href"].replace("../../", ""))
+            for link in label_data["links"]
+            if link["rel"] == "source"
+        ]
```
A lot of these list comprehensions would probably be more legible as for loops. I rarely use list comprehensions unless I can fit everything on one line.

KennSmithDS replied:
Swapped all the list compressions in `cloud_cover.py` to for loops, hope I didn't' break anything :-D

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,389 @@
+            suptitle: optional suptitle to use for figure
+
+        Returns:
+            a matplotlib Figure with the rendered sample
```
Should have a "Raises" section

KennSmithDS replied:
What would this likely raise? For the most part the plot method was copied using some of the other datasets as a template.

adamjstewart replied:
It raises ValueError a few lines later

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,389 @@
+        Returns:
+            a matplotlib Figure with the rendered sample
+
+        .. versionadded:: 0.2
```
This line can be removed, the entire class will be added in 0.4

adamjstewart reviewed (commented):
Thanks for the `data.py`, that looks like it wasn't easy to write!

Did a more detailed pass through the code and listed mostly minor things to change. I think this is pretty close to completion! Will likely have one last pass once these comments are addressed.

KennSmithDS commented on docs/api/datasets.rst:
```diff
@@ -152,6 +152,11 @@ BigEarthNet
+CloudCover
+^^^^^^^^^^
+
+.. autoclass:: CloudCoverDetection
```
This is my proposed line if I'm reading the headers correctly, although the existing entries don't see consistent with quotes or not:
`Cloud Cover Detection`_,S,"Radiant MLHub","11748,10980",2,512x512,10,MSI

adamjstewart replied:
Quotes are only needed when escaping commas. So this should work:
```rst
`Cloud Cover Detection`_,S,Radiant MLHub,"11,748","10,980",2,512x512,10,MSI
```

adamjstewart commented on docs/api/non_geo_datasets.csv:
```diff
@@ -2,6 +2,7 @@ Dataset,Task,Source,# Samples,# Classes,Size (px),Resolution (m),Bands
 `ADVANCE`_,C,"Google Earth, Freesound","5,075",13,512x512,0.5,RGB
 `Benin Cashew Plantations`_,S,Airbus Pléiades,70,6,"1,122x1,186",10,MSI
 `BigEarthNet`_,C,Sentinel-1/2,"590,326",19--43,120x120,10,"SAR, MSI"
+`Cloud Cover Detection`_,S,"Radiant MLHub","11748,10980",2,512x512,10,MSI
```
```suggestion
`Cloud Cover Detection`_,S,Radiant MLHub,"11,748","10,980",2,512x512,10,MSI
```

adamjstewart commented on docs/api/non_geo_datasets.csv:
```diff
@@ -2,6 +2,7 @@ Dataset,Task,Source,# Samples,# Classes,Size (px),Resolution (m),Bands
 `ADVANCE`_,C,"Google Earth, Freesound","5,075",13,512x512,0.5,RGB
 `Benin Cashew Plantations`_,S,Airbus Pléiades,70,6,"1,122x1,186",10,MSI
 `BigEarthNet`_,C,Sentinel-1/2,"590,326",19--43,120x120,10,"SAR, MSI"
+`Cloud Cover Detection`_,S,Radiant MLHub,"11,748","10,980",2,512x512,10,MSI
```
```suggestion
`Cloud Cover Detection`_,S,Sentinel-2,"22,728",2,512x512,10,MSI
```
Is it 22,728 in the entire dataset or just in the training data?

KennSmithDS replied:
I'm not clear on what the `# Samples` is supposed to represent. The entire dataset is 22,728, and the train/test split is what was provided in the CSV, as that's how the catalog is split.

adamjstewart replied:
It's supposed to be the total number, not the split. At least, that's what we've done for other datasets. Not all datasets have a train/test split.

adamjstewart commented on docs/api/non_geo_datasets.csv:
```diff
@@ -2,6 +2,7 @@ Dataset,Task,Source,# Samples,# Classes,Size (px),Resolution (m),Bands
 `ADVANCE`_,C,"Google Earth, Freesound","5,075",13,512x512,0.5,RGB
 `Benin Cashew Plantations`_,S,Airbus Pléiades,70,6,"1,122x1,186",10,MSI
 `BigEarthNet`_,C,Sentinel-1/2,"590,326",19--43,120x120,10,"SAR, MSI"
+`Cloud Cover Detection`_,S,Sentinel-2,"22,728",2,512x512,10,MSI
```
```suggestion
`CloudCover`_,S,Sentinel-2,"22,728",2,512x512,10,MSI
```
This name needs to match the name in `docs/api/datasets.rst`. Alternatively, you can change that name instead.

adamjstewart commented on docs/api/datasets.rst:
```diff
@@ -152,6 +152,11 @@ BigEarthNet
 
 .. autoclass:: BigEarthNet
 
+Cloud Cover Detection
+^^^^^^^^^^
```
```suggestion
Cloud Cover Detection
^^^^^^^^^^^^^^^^^^^^^
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,401 @@
+    `crowdsourcing competition
+    <https://www.drivendata.org/competitions/83/cloud-cover/>`_ on DrivenData.org, and
+    later on was validated using a team of expert annotators. See
+    `this website <https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1>`
```
```suggestion
    `this website <https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1>`_
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,401 @@
+    * 4 multi-spectral bands from Sentinel-2 L2A product. The four bands are
+      [B02, B03, B04, B08] (refer to Sentinel-2 documentation for more
+      information about the bands).
+    * Label raster for the corresponding source tile representing a binary classifcation
```
```suggestion
    * Label raster for the corresponding source tile representing a binary classification
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,401 @@
+      [B02, B03, B04, B08] (refer to Sentinel-2 documentation for more
+      information about the bands).
+    * Label raster for the corresponding source tile representing a binary classifcation
+      for if the pixel is is a cloud or not.
```
```suggestion
      for if the pixel is a cloud or not.
```

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,401 @@
+            if band not in self.band_names:
+                raise ValueError(f"'{band}' is an invalid band name.")
+
+    def _check_integrity(self) -> bool:
```
This is an optional request. I know you've already spent a ton of time on this PR, and I would rather it be merged as is than have you give up. With that said:

Our older datasets use `_check_integrity` and `_download`, but have a problem. If you have already downloaded the dataset but not yet extracted it, or if you have extracted it and deleted the tar/zip files, the dataset will try to re-download and re-extract everything. All of our newer datasets now have separate `_verify`, `_download`, and `_extract` methods to properly support this. We would like to move all of our datasets to the latter format at some point but haven't gotten around to it yet. 

So if you want to convert this dataset to the new format, that would help. But if you want to save that for another day or another person, I totally understand.

adamjstewart reviewed (commented):
Few typos and one optional request. Also, can you add an example figure created by `CloudCoverDetection.plot()` to the PR description? That would help make sure that plotting works correctly (and make the PR and new dataset look even more exciting).

calebrob6 commented:
Just reviewed and it looks great! Awesome job @KennSmithDS! This `data.py` is probably our most intricate yet and will help enormously once we move to pure pytest generated data.

adamjstewart commented on torchgeo/datasets/cloud_cover.py:
```diff
@@ -0,0 +1,401 @@
+    `crowdsourcing competition
+    <https://www.drivendata.org/competitions/83/cloud-cover/>`_ on DrivenData.org, and
+    later on was validated using a team of expert annotators. See
+    `this website <https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1>`_
```
```suggestion
    `this website <https://mlhub.earth/data/ref_cloud_cover_detection_challenge_v1>`__
```

calebrob6 commented:
Downloading isn't working for me. It seems that `utils.download_radiant_mlhub_collection(...)` isn't actually grabbing the tar files.

E.g. running
```
from torchgeo.datasets import CloudCoverDetection
ds = CloudCoverDetection(download=True, api_key="wouldn't you like to know")
```

I get:
```
File ~/code/torchgeo_kennsmith/torchgeo/datasets/cloud_cover.py:130, in CloudCoverDetection.__init__(self, root, split, bands, transforms, download, api_key, checksum)
    127 self.bands = bands
    129 if download:
--> 130     self._download(api_key)
    132 if not self._check_integrity():
    133     raise RuntimeError(
    134         "Dataset not found or corrupted. "
    135         + "You can use download=True to download it"
    136     )

File ~/code/torchgeo_kennsmith/torchgeo/datasets/cloud_cover.py:344, in CloudCoverDetection._download(self, api_key)
    340 target_archive_path = os.path.join(
    341     self.root, self.target_meta[self.split]["filename"]
    342 )
    343 for fn in [image_archive_path, target_archive_path]:
--> 344     extract_archive(fn, self.root)

File ~/code/torchgeo_kennsmith/torchgeo/datasets/utils.py:124, in extract_archive(src, dst)
    122 for suffix, extractor in suffix_and_extractor:
    123     if src.endswith(suffix):
--> 124         with extractor(src, "r") as f:
    125             f.extractall(dst)
    126         return

File /anaconda/envs/torchgeo/lib/python3.9/tarfile.py:1620, in TarFile.open(cls, name, mode, fileobj, bufsize, **kwargs)
   1618     saved_pos = fileobj.tell()
   1619 try:
-> 1620     return func(name, "r", fileobj, **kwargs)
   1621 except (ReadError, CompressionError):
   1622     if fileobj is not None:

File /anaconda/envs/torchgeo/lib/python3.9/tarfile.py:1684, in TarFile.gzopen(cls, name, mode, fileobj, compresslevel, **kwargs)
   1681     raise CompressionError("gzip module is not available")
   1683 try:
-> 1684     fileobj = GzipFile(name, mode + "b", compresslevel, fileobj)
   1685 except OSError:
   1686     if fileobj is not None and mode == 'r':

File /anaconda/envs/torchgeo/lib/python3.9/gzip.py:173, in GzipFile.__init__(self, filename, mode, compresslevel, fileobj, mtime)
    171     mode += 'b'
    172 if fileobj is None:
--> 173     fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
    174 if filename is None:
    175     filename = getattr(fileobj, 'name', '')

FileNotFoundError: [Errno 2] No such file or directory: 'data/ref_cloud_cover_detection_challenge_v1_train_source.tar.gz'
```

KennSmithDS commented:
RE: https://github.com/microsoft/torchgeo/pull/510#issuecomment-1205762120

@calebrob6 thanks for brining this to my attention. We released a new version of our `radiant_mlhub` Python client (released on June 1st) that adds ability to make spatial and temporal queries on the assets before downloading them, but one major change I just realized is that it no longer downloads the raw asset archives for each collection anymore. First it downloads the STAC only catalog archive, and then it filters and downloads the assets from the Catalog. This would complicate the checksum integrity work as well. Any datasets in Torchgeo that use the `radiant_mlhub` client will probably have to be updated, including `CloudCoverDetection`. In the future, we might also stop creating the huge `.tar.gz` archives in our MLHub dataset workflow due to download constraints when file sizes get too large. 

I think these Datasets might all be affected since that update is the probable root cause:

- `BeninSmallHolderCashews`
- `CloudCoverDetection`
- `CV4AKenyaCropType`
- `TropicalCycloneWindEstimation`
- `NASAMarineDebris`
- `SpaceNet`

Also unrelated to this issue, I don't know if it's an issue with my Apple MBP M1 chipset conflicting with the Torchgeo dependencies, but I recently rebuilt an environment from scratch using the [installation instructions ](https://torchgeo.readthedocs.io/en/latest/user/installation.html) to generate the plot and debug this download issue, but my Jupyter kernels keep crashing on the import statements. 

Any thoughts what might cause this or recommendations for preventing kernel failure in the torchgeo environment?

adamjstewart commented:
I also found that Jupyter didn't work for me on M1: https://github.com/jupyter/notebook/issues/6435

adamjstewart commented on torchgeo/datasets/__init__.py:
```diff
@@ -123,6 +124,7 @@
     "ChesapeakeVA",
     "ChesapeakeWV",
     "ChesapeakeCVPR",
+    "CloudCoverDetection",
```
This line should probably be down in the NonGeoDataset section

adamjstewart replied:
Can you move this line to the section below?

adamjstewart reviewed (approved):
So is this PR ready to go as long as we use an older version of radiant-mlhub? We can convert all datasets to support the newer radiant-mlhub in a follow-up PR.

adamjstewart commented:
Rebased to fix coverage. Just need to see an example plot and this should be ready to merge.

KennSmithDS commented:
@adamjstewart sorry for delay, we've had some high priority projects past 1-2 weeks that have required my attention.

1. Do you have an example of what the plot should look like, or will this do?

![cloud_cover_plot_idx_1](https://user-images.githubusercontent.com/30271891/187300548-db63dc45-d72a-4352-b094-0620fbaad365.png)

2. Is there a standard pixel value normalization for Sentinel 2? 

I just modified a single line of code in the `.plot(...)` method to divide the pixel values by 3000, which is arbitrary but also used in `BeninSmallHolderCashews`, another MLHub `NonGeoDataset` which has the added numpy dimension for time series in the data. The `np.clip(..)` method is also used while normalizing the S2 pixel values.

I'll have to push 1 more commit with the slight modification to the `.plot(...)` method after I hear back from you. Cheers!

adamjstewart commented:
1. Plot looks good, just wanted to make sure the image and label aren't flipped or something like that. You'd be surprised how often that happens.
2. No preference on normalization constant. I used 2000 for Sentinel2 because 3000 made the image a bit too dark. I don't know of a more rigorous way to define how Sentinel imagery should be visualized. You would think there would be an "official" way to plot a true color image...
